lib/kafka_producer.js

```diff
@@ -91,18 +111,82 @@ KafkaProducer.prototype.connect = function connect(onConnect) {
 
 KafkaProducer.prototype.produce = function produce(topic, message, timeStamp, callback) {
     var self = this;
+
     if (self.restClient) {
-        var produceMessage = self.getProduceMessage(topic, message, timeStamp, 'binary');
-        self.restClient.produce(produceMessage, function handleResponse(err, res) {
-            if (callback) {
-                callback(err, res);
-            }
-        });
+        if (self.batching) {
+            self.batch(topic, message, timeStamp, callback);
+        } else {
+            var produceMessage = self.getProduceMessage(topic, message, timeStamp, 'binary');
+            self.restClient.produce(produceMessage, function handleResponse(err, res) {
+                if (callback) {
+                    callback(err, res);
+                }
+            });
+        }
     } else if (callback) {
         callback(new Error('Kafka Rest Client is not initialized!'));
     }
 };
 
+// Add message to topic's BatchMessage in cache or add new topic to cache
+// Flush topic's BatchMessage on flushCycleSecs interval or if greater than maxBatchSizeBytes
+KafkaProducer.prototype.batch = function batch(topic, message, timeStamp, callback) {
+    var self = this;
+    var messageBatch;
+
+    if (self.topicToBatchQueue[topic]) {
+        messageBatch = self.topicToBatchQueue[topic];
+        if ((messageBatch.sizeBytes + Buffer.byteLength(message)) > self.maxBatchSizeBytes) {
```

Comment:
Need a `messageBatch.numMessages > 0` gaurd here. We shouldnt flush an empty batch.